src/main/java/appeng/parts/p2p/PartP2PFEPower.java

```diff
@@ -73,180 +65,156 @@ public void onTunnelNetworkChange()
+		{
+			return 0;
+		}
 
-	private IEnergyStorage getOutput()
-	{
-		if( this.isOutput() )
+		@Override
+		public int receiveEnergy( int maxReceive, boolean simulate )
 		{
-			if( !this.cachedTarget )
+			if( PartP2PFEPower.this.isActive() )
 			{
-				final TileEntity self = this.getTile();
-				final TileEntity te = self.getWorld().getTileEntity( new BlockPos( self.getPos().getX() + this.getSide().xOffset, self.getPos()
-						.getY() + this.getSide().yOffset, self.getPos().getZ() + this.getSide().zOffset ) );
+				int total = 0;
 
-				if( te != null && te.hasCapability( Capabilities.FORGE_ENERGY, this.getSide().getOpposite().getFacing() ) )
+				try
 				{
-					this.outputTarget = te.getCapability( Capabilities.FORGE_ENERGY, this.getSide().getOpposite().getFacing() );
+					final int outputTunnels = PartP2PFEPower.this.getOutputs().size();
+
+					if( outputTunnels == 0 )
+					{
+						return 0;
+					}
+
+					final int amountPerOutput = maxReceive / outputTunnels;
+					int overflow = maxReceive % amountPerOutput;
+
+					for( PartP2PFEPower target : PartP2PFEPower.this.getOutputs() )
+					{
+						final IEnergyStorage output = target.getOutput();
+						final int toSend = amountPerOutput + overflow;
+						final int received = output.receiveEnergy( toSend, simulate );
+
+						overflow = toSend - received;
+						total += received;
+					}
+
+					PartP2PFEPower.this.queueTunnelDrain( PowerUnits.RF, total );
 				}
-				else
+				catch( GridAccessException ignored )
 				{
-					this.outputTarget = null;
 				}
 
-				this.cachedTarget = true;
+				return total;
 			}
 
-			if( this.outputTarget != null && this.outputTarget.canReceive() )
-			{
-				return this.outputTarget;
-			}
-		}
-
-		return NULL_ENERGY_STORAGE;
-	}
-
-	@Override
-	public int getEnergyStored()
-	{
-		if( this.isOutput() || !this.isActive() )
-		{
 			return 0;
 		}
 
-		int total = 0;
-
-		try
-		{
-			for( PartP2PFEPower t : this.getOutputs() )
-			{
-				total += t.getOutput().getEnergyStored();
-			}
-		}
-		catch( GridAccessException e )
+		@Override
+		public boolean canExtract()
 		{
-			return 0;
+			return false;
 		}
 
-		return total;
-	}
-
-	@Override
-	public int getMaxEnergyStored()
-	{
-		if( this.isOutput() || !this.isActive() )
+		@Override
+		public boolean canReceive()
 		{
-			return 0;
+			return true;
 		}
 
-		int total = 0;
-
-		try
+		@Override
+		public int getMaxEnergyStored()
 		{
-			for( PartP2PFEPower t : this.getOutputs() )
+			if( !PartP2PFEPower.this.isActive() )
 			{
-				total += t.getOutput().getMaxEnergyStored();
+				return 0;
 			}
-		}
-		catch( GridAccessException e )
-		{
-			return 0;
-		}
 
-		return total;
-	}
+			int total = 0;
 
-	@Override
-	public boolean canExtract()
-	{
-		return false;
-	}
+			try
+			{
+				for( PartP2PFEPower t : PartP2PFEPower.this.getOutputs() )
+				{
+					total += t.getOutput().getMaxEnergyStored();
+				}
+			}
+			catch( GridAccessException e )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)